src/castling.rs

```diff
@@ -0,0 +1,178 @@
+                },
+                pos::h1 => Castling {
+                    white: CastleAbility {
+                        king: false,
+                        queen: self.white.queen,
+                    },
+                    ..self
+                },
+                pos::a1 => Castling {
+                    white: CastleAbility {
+                        king: self.white.king,
+                        queen: false,
+                    },
+                    ..self
+                },
+                _ => self,
+            },
+            Player::Black => match pos {
+                pos::e8 => Castling {
+                    black: CastleAbility {
+                        king: false,
+                        queen: false,
+                    },
+                    ..self
+                },
+                pos::h8 => Castling {
+                    black: CastleAbility {
+                        king: false,
+                        queen: self.black.queen,
+                    },
+                    ..self
+                },
+                pos::a8 => Castling {
+                    black: CastleAbility {
+                        king: self.black.king,
+                        queen: false,
+                    },
+                    ..self
+                },
+                _ => self,
+            },
+        }
+    }
+
+    /// Castle. Returns the new castling and board state.
+    pub fn castle(self, board: &Board, player: Player, kingside: bool) -> (Board, Self) {
+        if player.is_white() {
+            let next_castling = Castling {
+                white: CastleAbility {
+                    king: false,
+                    queen: false,
+                },
+                ..self
+            };
+            let next_board = if kingside {
+                board
+                    .move_piece(pos::e1, pos::g1)
+                    .move_piece(pos::h1, pos::f1)
+            } else {
+                board
+                    .move_piece(pos::e1, pos::c1)
+                    .move_piece(pos::a1, pos::d1)
+            };
+            (next_board, next_castling)
+        } else {
+            let next_castling = Castling {
+                black: CastleAbility {
+                    king: false,
+                    queen: false,
+                },
+                ..self
+            };
+            let next_board = if kingside {
+                board
+                    .move_piece(pos::e8, pos::g8)
+                    .move_piece(pos::h8, pos::f8)
+            } else {
+                board
+                    .move_piece(pos::e8, pos::c8)
+                    .move_piece(pos::a8, pos::d8)
+            };
+            (next_board, next_castling)
+        }
+    }
+
+    /// Is the castling for `player` unobstructed at `kingside` on a given `board`?
+    pub fn free(board: &Board, player: Player, kingside: bool) -> bool {
+        match (player, kingside) {
+            (Player::White, true) => {
```

Comment:
Actually you might be able to avoid matching on each player by making associated constant positions for each player. E.g. `Player::White` could have an associated constant for the home of its king so you would never need to write out `d1` explicitly.

Reply:
we could, but then i still need the correct positions of the rook and the target squares, so i'd have to put those constants somewhere. 

i ended up cleaning this up a bit with adding a "without" function: d2d8de8993e0d3a5c37f935b2fcae049030bc424